File: android/src/main/kotlin/com/neelakandan/flutter_neo_shield/location/SensorFusionValidator.kt
package com.neelakandan.flutter_neo_shield.location

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import kotlin.math.abs
import kotlin.math.sqrt

/**
 * Layer 5: Sensor Fusion Validation.
 *
 * Cross-correlates GPS movement with accelerometer/gyroscope data.
 * If GPS says moving but sensors show stationary → spoofed.
 */
class SensorFusionValidator(private val context: Context) : SensorEventListener {

    private val sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as? SensorManager
    private var lastAccelEnergy = 0.0f
    private var lastCorrelationScore = 0.0
    private var accelSamples = mutableListOf<FloatArray>()

    init {
        startSensorMonitoring()
    }

    fun getLastCorrelationScore(): Double = lastCorrelationScore

    private fun startSensorMonitoring() {
        try {
            val accel = sensorManager?.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
            if (accel != null) {
                sensorManager?.registerListener(this, accel, SensorManager.SENSOR_DELAY_NORMAL)
            }
        } catch (_: Exception) {}
    }

    override fun onSensorChanged(event: SensorEvent) {
        if (event.sensor.type == Sensor.TYPE_ACCELEROMETER) {
            accelSamples.add(event.values.copyOf())
            if (accelSamples.size > 50) {
                accelSamples.removeAt(0)
            }
            lastAccelEnergy = calculateAccelEnergy(accelSamples)
        }
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}

    /** Check if sensor availability is suspicious (emulator/spoofed env). */
    fun checkSensorAvailability(): Double {
        var score = 0.0
        if (sensorManager?.getDefaultSensor(Sensor.TYPE_ACCELEROMETER) == null) score += 0.2
        if (sensorManager?.getDefaultSensor(Sensor.TYPE_GYROSCOPE) == null) score += 0.2
        if (sensorManager?.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD) == null) score += 0.1
        if (sensorManager?.getDefaultSensor(Sensor.TYPE_PRESSURE) == null) score += 0.1
        return score.coerceIn(0.0, 1.0)
    }

    /** Correlate GPS speed with accelerometer energy.
     * Call this with the GPS-reported speed to update correlation score. */
    fun updateWithGpsSpeed(gpsSpeedMs: Float) {
        var score = 0.0

        // GPS shows fast movement but no accelerometer activity
        if (gpsSpeedMs > 5.0f && lastAccelEnergy < 0.2f) {
            score += 0.7
        }

        // GPS stationary but significant accelerometer activity
        if (gpsSpeedMs < 0.5f && lastAccelEnergy > 3.0f) {
            score += 0.3
        }

        // Sensor unavailability adds suspicion
        score += checkSensorAvailability() * 0.3

        lastCorrelationScore = score.coerceIn(0.0, 1.0)
    }

    private fun calculateAccelEnergy(samples: List<FloatArray>): Float {
        if (samples.isEmpty()) return 0.0f
        var sumSq = 0.0f
        for (sample in samples) {
            val magnitude = sqrt(
                sample[0] * sample[0] + sample[1] * sample[1] + sample[2] * sample[2]
            )
            val deviation = magnitude - 9.81f
            sumSq += deviation * deviation
        }
        return sqrt(sumSq / samples.size)
    }

    fun dispose() {
        try {
            sensorManager?.unregisterListener(this)
        } catch (_: Exception) {}
    }
}

File: ios/Classes/location/SensorFusionValidator.swift
import CoreMotion
import CoreLocation

/// Layer 5: Sensor Fusion Validation for iOS.
class SensorFusionValidator {

    private let motionManager = CMMotionManager()
    private let activityManager = CMMotionActivityManager()
    private var lastCorrelationScore: Double = 0.0
    private var lastActivity: CMMotionActivity?
    private var isMonitoring = false

    init() {
        startMonitoring()
    }

    func getLastCorrelationScore() -> Double { return lastCorrelationScore }

    private func startMonitoring() {
        guard !isMonitoring else { return }
        isMonitoring = true

        if CMMotionActivityManager.isActivityAvailable() {
            activityManager.startActivityUpdates(to: .main) { [weak self] activity in
                self?.lastActivity = activity
            }
        }

        if motionManager.isAccelerometerAvailable {
            motionManager.accelerometerUpdateInterval = 0.1
            motionManager.startAccelerometerUpdates()
        }
    }

    /// Correlate GPS data with motion sensors.
    func correlateWithGPS(gpsSpeed: Double) -> Double {
        var score: Double = 0.0

        // Check activity vs GPS speed
        if let activity = lastActivity {
            // GPS says fast but activity says stationary
            if gpsSpeed > 5.0 && activity.stationary {
                score += 0.6
            }
            // GPS says automotive but activity says walking
            if gpsSpeed > 20.0 && activity.walking && !activity.automotive {
                score += 0.4
            }
        }

        // Check accelerometer energy vs GPS speed
        if let accelData = motionManager.accelerometerData {
            let accel = accelData.acceleration
            let magnitude = sqrt(accel.x * accel.x + accel.y * accel.y + accel.z * accel.z)
            let deviation = abs(magnitude - 1.0) // gravity = 1.0g

            // GPS fast but no acceleration
            if gpsSpeed > 5.0 && deviation < 0.02 {
                score += 0.5
            }
        }

        // Sensor availability check
        score += checkSensorAvailability() * 0.3

        lastCorrelationScore = min(score, 1.0)
        return lastCorrelationScore
    }

    func checkSensorAvailability() -> Double {
        var score: Double = 0.0
        if !motionManager.isAccelerometerAvailable { score += 0.2 }
        if !motionManager.isGyroAvailable { score += 0.2 }
        if !motionManager.isMagnetometerAvailable { score += 0.1 }
        if !CMAltimeter.isRelativeAltitudeAvailable() { score += 0.1 }
        if !CMMotionActivityManager.isActivityAvailable() { score += 0.1 }
        if !CMPedometer.isStepCountingAvailable() { score += 0.1 }
        return min(score, 1.0)
    }

    func dispose() {
        motionManager.stopAccelerometerUpdates()
        activityManager.stopActivityUpdates()
        isMonitoring = false
    }
}
